src/charm.py

```diff
@@ -63,6 +82,95 @@ def _on_website_relation_joined(self, event):
             'port': str(port)
         })
 
+    def _setup_promtail(self):
+        # Download promtail binary
+        check_call(["curl", "-LO", "https://github.com/grafana/loki/releases/download/v2.6.1/promtail-linux-amd64.zip"])
+        check_call(["busybox", "unzip", "promtail-linux-amd64.zip"])
+        check_call(["sudo", "mv", "promtail-linux-amd64", "/usr/local/bin/promtail"])
+        check_call(["sudo", "mkdir", "-p", "/etc/promtail"])
+
+        # Create promtail service
+        svc_file = open("/etc/systemd/system/promtail.service", "w")
+        svc_file.write('''[Unit]
+Description=Promtail service
+After=network.target
+
+[Service]
+Type=simple
+User=root
+ExecStart=/usr/local/bin/promtail -config.file /etc/promtail/config.yaml
+
+[Install]
+WantedBy=multi-user.target
+''')
+        svc_file.close()
+        check_call(["sudo", "systemctl", "daemon-reload"])
+
+        # Start promtail service
+        self._restart_promtail()
+
+    def _restart_promtail(self, _ = None):
+        '''Reconfigures and restarts the promtail service when the logging
+        relation is changed.
+        Assumes the promtail service has already been set up by _setup_promtail
+        '''
+
+        # Find client push urls
+        endpoints = []
+        for relation in self.model.relations[LOGGING_RELATION_NAME]:
+            for unit in relation.units:
+                ip = relation.data[unit].get("private-address")
```

Comment:
We should be moving away from private-address and looking to use network-info instead

Reply:
Here's a dump of `relation.data`:
```
{
  <ops.model.Unit loki-logging/0>: {
    'egress-subnets': '10.152.183.23/32',
    'endpoint': '{
      "url": "http://loki-0.loki-endpoints.offers.svc.cluster.local:3100/loki/api/v1/push"
    }',
    'ingress-address': '10.152.183.23',
    'private-address': '10.152.183.23',
    'public_address': 'loki-0.loki-endpoints.offers.svc.cluster.local'
  },
  <ops.model.Application loki-logging>: {
    'promtail_binary_zip_url': '{
      "amd64": {
        "filename": "promtail-static-amd64",
        "zipsha": "543e333b0184e14015a42c3c9e9e66d2464aaa66eca48b29e185a6a18f67ab6d",
        "binsha": "17e2e271e65f793a9fbe81eab887b941e9d680abe82d5a0602888c50f5e0cac9",
        "url": "https://github.com/canonical/loki-k8s-operator/releases/download/promtail-v2.5.0/promtail-static-amd64.gz"
      }
    }'
  }
}
```

Can't see anything there about network-info. What exactly do you mean?

Reply:
In any case, prefer `ingress-address`, which will yield the same result. `private-address` is deprecated.